### ISSUE TITLE
DependencySubstitutions API evolution (deprecate 'Substitution.with')

### ISSUE THREAD

**Issue by ljacomet** [gradle/gradle#14107]
Gradle 6.6 introduced `Substitution.using` as an alternative to `Substitution.with` which return the `Substitution` and enables chaining.
We should de-incubate `using` and deprecate `with` in Gradle 7

**Comment by jjohannes** [comment]
`using` is promoted for 7.0 as part of #15650.

**Comment by rieske** [comment]
For now, we can only add the deprecation annotations on Java methods and documentation, but can not nag when this deprecated method is used, because it is used by the Android plugin and nagging here causes a significant performance regression.
https://github.com/gradle/gradle/pull/16731#discussion_r609705836

Leaving this open for now and let's revisit once a new version of Android plugin is released.

**Comment by rieske** [comment]
Because there are more instances of newly deprecated method usages in Android plugin, I created a separate issue to follow up on them and enable nagging: https://github.com/gradle/gradle/issues/16782

This one can be closed once https://github.com/gradle/gradle/pull/16731 is merged.
